processing/src/main/java/org/apache/druid/query/aggregation/AbstractSerializablePairSerde.java

```diff
@@ -0,0 +1,116 @@
+ * under the License.
+ */
+
+package org.apache.druid.query.aggregation;
+
+import com.google.common.primitives.Longs;
+import org.apache.druid.collections.SerializablePair;
+import org.apache.druid.data.input.InputRow;
+import org.apache.druid.segment.GenericColumnSerializer;
+import org.apache.druid.segment.column.ColumnBuilder;
+import org.apache.druid.segment.data.GenericIndexed;
+import org.apache.druid.segment.data.ObjectStrategy;
+import org.apache.druid.segment.serde.ComplexColumnPartSupplier;
+import org.apache.druid.segment.serde.ComplexMetricExtractor;
+import org.apache.druid.segment.serde.ComplexMetricSerde;
+import org.apache.druid.segment.serde.LargeColumnSupportedComplexColumnSerializer;
+import org.apache.druid.segment.writeout.SegmentWriteOutMedium;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+
+/**
+ * The class serializes a Pair<Long, ?> object for double/float/longFirst and double/float/longLast aggregators
+ */
+public abstract class AbstractSerializablePairSerde<T extends SerializablePair<Long, ?>> extends ComplexMetricSerde
+{
+  private final Class<T> pairClassObject;
+
+  public AbstractSerializablePairSerde(Class<T> pairClassObject)
+  {
+    this.pairClassObject = pairClassObject;
+  }
+
+  @Override
+  public ComplexMetricExtractor getExtractor()
+  {
+    return new ComplexMetricExtractor()
+    {
+      @Override
+      public Class<T> extractedClass()
+      {
+        return pairClassObject;
+      }
+
+      @Override
+      public Object extractValue(InputRow inputRow, String metricName)
+      {
+        return inputRow.getRaw(metricName);
+      }
+    };
+  }
+
+  @Override
+  public void deserializeColumn(ByteBuffer buffer, ColumnBuilder columnBuilder)
+  {
+    final GenericIndexed column = GenericIndexed.read(buffer, getObjectStrategy(), columnBuilder.getFileMapper());
+    columnBuilder.setComplexColumnSupplier(new ComplexColumnPartSupplier(getTypeName(), column));
+  }
+
+  @Override
+  public ObjectStrategy<T> getObjectStrategy()
+  {
+    return new ObjectStrategy<T>()
+    {
+      @Override
+      public int compare(@Nullable T o1, @Nullable T o2)
+      {
+        return Longs.compare(o1.lhs, o2.lhs);
+      }
+
+      @Override
+      public Class<T> getClazz()
+      {
+        return pairClassObject;
+      }
+
+      @Override
+      public T fromByteBuffer(ByteBuffer buffer, int numBytes)
+      {
+        return toPairObject(buffer);
+      }
+
+      @Override
+      public byte[] toBytes(T val)
+      {
+        return pairToBytes(val);
+      }
+    };
+  }
+
+  @Override
+  public GenericColumnSerializer<T> getSerializer(SegmentWriteOutMedium segmentWriteOutMedium, String column)
+  {
+    return LargeColumnSupportedComplexColumnSerializer.create(segmentWriteOutMedium, column, this.getObjectStrategy());
+  }
+
+  protected abstract T toPairObject(ByteBuffer buffer);
+
+  protected abstract byte[] pairToBytes(T val);
```

Comment:
```suggestion
  protected abstract byte[] pairToBytes(@Nullable T val);
```

Reply:
In the new commit, null check is added to the caller of this method, so it's not necessary to declare this parameter as `Nullable`